File: 105-construct-binary-tree-from-preorder-and-inorder-traversal/construct-binary-tree-from-preorder-and-inorder-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* build(vector<int>& preorder, vector<int>& inorder, int& root, int l, int r)
    {
        if(l>r)
           return NULL;
        if(root>=preorder.size())
           return NULL;
       int x = preorder[root];
       int j;
       for(j=l; j<=r; j++)
       {
           if(inorder[j]==x)
              break;
       }
       TreeNode* rt = new TreeNode(x);
       root++;
       rt->left = build(preorder, inorder, root, l, j-1);
       rt->right = build(preorder, inorder, root, j+1, r);
       return rt;
    }
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        int root=0;
        return build(preorder, inorder, root, 0, inorder.size()-1);
    }
};

File: 105-construct-binary-tree-from-preorder-and-inorder-traversal/construct-binary-tree-from-preorder-and-inorder-traversal.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    public TreeNode calc(Dictionary<int, int> Inorder, int[] preorder, int l, int r, ref int ind)
    {
    int n = preorder.Count();
    if(ind==n || l>r)
        return null;
    else
    {
        int pos = Inorder[preorder[ind]];
        TreeNode root=new TreeNode();
        root.val=preorder[ind];
        ind++;
        root.left=calc(Inorder, preorder, l, pos-1, ref ind);
        root.right=calc(Inorder, preorder, pos+1, r, ref ind);
        return root;
    }
    }
    public TreeNode BuildTree(int[] preorder, int[] inorder) {
        Dictionary<int, int> Inorder = new Dictionary<int, int>();
        int n = inorder.Count();

        for(int i=0; i<n; i++)
        {
           Inorder[inorder[i]]=i;
        }
        int ind=0;
        return calc(Inorder, preorder, 0, n-1, ref ind);
    }
}